### PR TITLE
feat: Add option `schemaCacheTtl` for schema cache pulling as alternative to `enableSchemaHooks`

### LINKED ISSUE #8229: Re-add schema pulling
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
[Parse Server 5](https://github.com/parse-community/parse-server/releases/tag/5.0.0) introduced schema hooks with the option `databaseOptions.enableSchemaHooks` in https://github.com/parse-community/parse-server/issues/7214. There are potential performance downsides using the MongoDB change stream.

The MongoDB change stream works by setting a cursor on the MongoDB oplog collection. If many Parse Server instances are behind a load balancer, each one will open a change stream and set an oplog cursor. Each cursor adds an overhead to the MongoDB cluster resources in terms of CPU and RAM consumption.

There may be deployments in which schema hooks are disadvantageous. If the schema doesn't change much, it's unnecessary to constantly maintain the database cursors and the previous solution of pulling the schema in regular intervals may be much more economic. In [hindsight](https://github.com/parse-community/parse-server/pull/7214#issuecomment-810995869), it seems that schema hooks should have been added as a optional feature, not a mandatory change.

### Steps to reproduce

Run many server instances in a cluster with high write frequency and watch the MongoDB resource consumption increase significantly.

### Actual Outcome

The developer has no choice and is forced to use schema hooks, or restart all server instances on every schema change.

### Expected Outcome

The developer should have the choice of pulling the schema in defined intervals, as was possible with Parse Server <5.

### Workaround

Disable `databaseOptions.enableSchemaHooks` and restart server instances whenever schema changed.

### Environment

Server
- Parse Server version: `5.7.2`

### PR DESCRIPTION
**Pull request by dblythy**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue

If a developer is not using `enableSchemaHooks`, the server will have to be restarted to fetch new schema data

Closes: #8229

## Approach

Adds a conditional reload of schema if it hasn't been refreshed below `databaseOptions.schemaCacheTTL`

## Tasks

- [x] Add tests

### CODE REVIEW

**Comment by mtrezza** [comment]
I think it would be good to make this TTL refresh optional. So if the option is not set, there should be no refresh happening, which is the status quo. This way it's also less of breaking change.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 1]
```diff
@@ -139,11 +139,12 @@ export class MongoStorageAdapter implements StorageAdapter {
   _maxTimeMS: ?number;
   canSortOnJoinTables: boolean;
   enableSchemaHooks: boolean;
+  schemaCacheTTL: ?number;
```
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Use camel case `schemaCacheTtl`.

**Review by mtrezza — CHANGES_REQUESTED** [review]
What happens if `schemaCacheTtl` and `enableSchemaHooks` are set? Should both be enabled? There may be a use case for it, who knows. Or should it throw an error or a warning?

**Comment by dblythy** [comment]
I think if they are both enabled:

- Schema will update when change event is triggered
- Schema will also pull every `ttl`

Perhaps if the network connection between the Parse Server and the database is inconsistent, `change` events might be missed.

**Comment by mtrezza** [comment]
Makes sense, so let's just allow both features to be configured independently without any warning

File: src/Options/index.js [hunk 2]
```diff
@@ -548,6 +548,8 @@ export interface DatabaseOptions {
   /* Enables database real-time hooks to update single schema cache. Set to `true` if using multiple Parse Servers instances connected to the same database. Failing to do so will cause a schema change to not propagate to all instances and re-syncing will only happen when the instances restart. To use this feature with MongoDB, a replica set cluster with [change stream](https://docs.mongodb.com/manual/changeStreams/#availability) support is required.
   :DEFAULT: false */
   enableSchemaHooks: ?boolean;
+  /* The TTL for caching the schema for optimizing read/write operations. You should put a long TTL when your DB is in production */
```
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires. */
```

File: src/Options/docs.js [hunk 3]
```diff
@@ -225,6 +225,7 @@
 /**
  * @interface DatabaseOptions
  * @property {Boolean} enableSchemaHooks Enables database real-time hooks to update single schema cache. Set to `true` if using multiple Parse Servers instances connected to the same database. Failing to do so will cause a schema change to not propagate to all instances and re-syncing will only happen when the instances restart. To use this feature with MongoDB, a replica set cluster with [change stream](https://docs.mongodb.com/manual/changeStreams/#availability) support is required.
+ * @property {Number} schemaCacheTtl The TTL for caching the schema for optimizing read/write operations. You should put a long TTL when your DB is in production
```
Comment (mtrezza) [inline src/Options/docs.js]:
```suggestion
 * @property {Number} schemaCacheTtl The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires.
```

File: src/Options/Definitions.js [hunk 4]
```diff
@@ -971,6 +971,12 @@ module.exports.DatabaseOptions = {
     action: parsers.booleanParser,
     default: false,
   },
+  schemaCacheTtl: {
+    env: 'PARSE_SERVER_DATABASE_SCHEMA_CACHE_TTL',
+    help:
+      'The TTL for caching the schema for optimizing read/write operations. You should put a long TTL when your DB is in production',
```
Comment (mtrezza) [inline src/Options/Definitions.js]:
```suggestion
      'The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires.',
```

**Review by mtrezza — CHANGES_REQUESTED** [review]
Adding doc

**Review by mtrezza — DISMISSED** [review]
Looks good, waiting for CI...

File: src/Options/index.js [hunk 5]
```diff
@@ -548,6 +548,8 @@ export interface DatabaseOptions {
   /* Enables database real-time hooks to update single schema cache. Set to `true` if using multiple Parse Servers instances connected to the same database. Failing to do so will cause a schema change to not propagate to all instances and re-syncing will only happen when the instances restart. To use this feature with MongoDB, a replica set cluster with [change stream](https://docs.mongodb.com/manual/changeStreams/#availability) support is required.
   :DEFAULT: false */
   enableSchemaHooks: ?boolean;
+  /* The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires. */
```
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. Default is infinite which means the schema cache never expires. */
```

File: src/Options/docs.js [hunk 6]
```diff
@@ -225,6 +225,7 @@
 /**
  * @interface DatabaseOptions
  * @property {Boolean} enableSchemaHooks Enables database real-time hooks to update single schema cache. Set to `true` if using multiple Parse Servers instances connected to the same database. Failing to do so will cause a schema change to not propagate to all instances and re-syncing will only happen when the instances restart. To use this feature with MongoDB, a replica set cluster with [change stream](https://docs.mongodb.com/manual/changeStreams/#availability) support is required.
+ * @property {Number} schemaCacheTtl The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires.
```
Comment (mtrezza) [inline src/Options/docs.js]:
```suggestion
 * @property {Number} schemaCacheTtl The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. Default is infinite which means the schema cache never expires.
```

File: src/Options/Definitions.js [hunk 7]
```diff
@@ -971,6 +971,12 @@ module.exports.DatabaseOptions = {
     action: parsers.booleanParser,
     default: false,
   },
+  schemaCacheTtl: {
+    env: 'PARSE_SERVER_DATABASE_SCHEMA_CACHE_TTL',
+    help:
+      'The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. By default this option is disabled and the schema cache never expires.',
```
Comment (mtrezza) [inline src/Options/Definitions.js]:
```suggestion
      'The duration in seconds after which the schema cache expires and will be refetched from the database. Use this option if using multiple Parse Servers instances connected to the same database. A low duration will cause the schema cache to be updated too often, causing unnecessary database reads. A high duration will cause the schema to be updated too rarely, increasing the time required until schema changes propagate to all server instances. This feature can be used as an alternative or in conjunction with the option `enableSchemaHooks`. Default is infinite which means the schema cache never expires.',
```

**Review by mtrezza — COMMENTED** [review]
Doc rewording

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.35](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.35)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/6.1.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0](https://github.com/parse-community/parse-server/releases/tag/6.1.0)

**Pull request merged**